src/utils/url.ts

```diff
@@ -0,0 +1,13 @@
+export function normalizeProtocolUrl(url: string): string {
+  const fullUrl = url.includes('://') ? url : `http://${url}`;
+  const { hostname, pathname } = new URL(fullUrl);
```

Comment:
why not include the `search` as well?  i feel like that's a pretty commonly used way of distinguishing resources on many webservers

NIT: we may want to `try { ... } catch { }` this just in case

Reply:
Good call on try/catch. will add that today.

Excluding `search` because there is some conversation lately around how to configure different "roles" of a protocol. For example, for TBDex a DWN may want to set up as an "issuer" or as a "PFI" or a "wallet". Each of those roles would use the TBDex protocol a bit differently. The current idea that has been tossed around DWN office hours is to have that "role" be specified in a query param. So they should be normalized out.

This is a very very WIP idea, so we may end up changing it in the coming months (or even weeks), so if keeping `search` ends up being the right approach, I'm open to changing it later.

Reply:
hmmm if that's the case then how would someone query for a specific "role" of the same protocol?

since we're always normalizing the `protocol` anywhere we send it to `MessageStore` (and therefore `IndexLevel`), would it be up to the caller to do their own filtering?

Reply:
ah sorry i shouldve been clearer.  i was specifically referring to using `RecordsQuery` to something like

> get all messages with `{ protocol: 'example.com?role=issuer' }`

AFAICT what you currently have in this PR would return everything matching `{ protocol: 'example.com' }`, which could include other things (e.g. `{ protocol: 'example.com?role=other' }`) that the caller might not expect

we may want to make the normalization even more minimal than what's there now, such as only

- always having a `protocol` of `http://`
- removing trailing `'/'` from the `pathname`
- removing `port`
- removing `password`
- sorting the `search`
- removing `hash`

IMO this is part of why i dont think it's a good idea to do normalization, as we're effectively constraining the possibilities of URLs without any sort of specification as to what those constraints are (e.g. it's totally valid for a different port to return a different result).  i understand that we're trying to make the developer experience better by trying to avoid (possibly) common mistakes (e.g. adding a trailing `'/'`), but i think that could just result in developers being more confused by seemingly magical stuff happening instead of exactly what they asked for

---

as far as what you said, i think so long as `example.com?role=issuer` and `example.com?role=pfi` resolve to different things then that's fine :)

Reply:
I'm starting to agree with you as I jam on edge cases. Lemme think about this a bit